src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java

```diff
@@ -1337,159 +1369,402 @@ public void setMirror()
+     * @param requestToken The {@link IToken} that is used to represent the request.
+     * @param requested    The class of the type that has been requested eg. {@code ItemStack.class}
      */
-    public static class View
+    private void addRequestToMaps(@NotNull Integer citizenId, @NotNull IToken requestToken, @NotNull TypeToken requested)
     {
-        private final ColonyView colony;
-        @NotNull
-        private final BlockPos   location;
+        if (!openRequests.containsKey(requested))
+        {
+            openRequests.put(requested, new ArrayList<>());
+        }
+        openRequests.get(requested).add(requestToken);
 
-        private int buildingLevel    = 0;
-        private int buildingMaxLevel = 0;
-        private int workOrderLevel   = NO_WORK_ORDER;
+        requestsByCitizen.put(requestToken, citizenId);
 
-        /**
-         * Creates a building view.
-         *
-         * @param c ColonyView the building is in.
-         * @param l The location of the building.
-         */
-        protected View(final ColonyView c, @NotNull final BlockPos l)
+        if (!citizensByRequests.containsKey(citizenId))
         {
-            colony = c;
-            location = new BlockPos(l);
+            citizensByRequests.put(citizenId, new ArrayList<>());
         }
+        citizensByRequests.get(citizenId).add(requestToken);
+    }
 
-        /**
-         * Gets the id for this building.
-         *
-         * @return A BlockPos because the building ID is its location.
-         */
-        @NotNull
-        public BlockPos getID()
+    public boolean hasWorkerOpenRequests(@NotNull CitizenData citizen)
+    {
+        return !getOpenRequests(citizen).isEmpty();
+    }
+
+    public ImmutableList<IRequest> getOpenRequests(@NotNull final CitizenData data)
+    {
+        if (!citizensByRequests.containsKey(data.getId()))
         {
-            // Location doubles as ID
-            return location;
+            return ImmutableList.of();
         }
 
-        /**
-         * Gets the location of this building.
-         *
-         * @return A BlockPos, where this building is.
-         */
-        @NotNull
-        public BlockPos getLocation()
+        return ImmutableList.copyOf(citizensByRequests.get(data.getId()).stream().map(getColony().getRequestManager()::getRequestForToken).filter(Objects::nonNull).iterator());
+    }
+
+    public boolean hasWorkerOpenRequestsFiltered(@NotNull CitizenData citizen, @NotNull Predicate<IRequest> selectionPredicate)
+    {
+        return getOpenRequests(citizen).stream().anyMatch(selectionPredicate);
+    }
+
+    public <R> boolean hasWorkerOpenRequestsOfType(@NotNull final CitizenData citizenData, final TypeToken<R> requestType)
+    {
+        return !getOpenRequestsOfType(citizenData, requestType).isEmpty();
+    }
+
+    @SuppressWarnings(Suppression.GENERIC_WILDCARD)
+    public <R> ImmutableList<IRequest<? extends R>> getOpenRequestsOfType(
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Remove usage of generic wildcard type. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)